### PR TITLE
SWIFT-36, 37: Implement idiomatic support for command monitoring and corresponding spec tests

### PR DESCRIPTION
**Pull request by kmahar**
(This uses the changes from the other PR regarding integers)

This adds support for command monitoring via `Notification`s and the default `NotificationCenter`. Also implements the corresponding spec tests. I encountered a number of odd edge cases there which I have attempted to document

### CODE REVIEW

File: Tests/MongoSwiftTests/CommandMonitoringTests.swift [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+
+                // 1. Setup the specified DB and collection with provided data
+                let db = try client.db(testFile.databaseName)
+                try db.drop() // In case last test run failed, drop to clear out DB
+                let collection = try db.createCollection(testFile.collectionName)
+                _ = try collection.insertMany(testFile.data)
+
+                // 2. Add an observer that looks for all events
+                var expectedEvents = test.expectations
+                let observer = center.addObserver(forName: nil, object: nil, queue: nil) { (notif) in
```
Comment (kmahar) [inline Tests/MongoSwiftTests/CommandMonitoringTests.swift]: FYI, `forName` could be `.commandSucceeded`, `.commandStarted`, etc. if you only want particular named events

File: Sources/MongoSwift/APM.swift [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+     *  If no specific event types are provided, all events will be posted. 
+     *
+     *  Calling this function will reset all previously enabled events - i.e.
+     *  calling
+     *      client.enableNotifications(forEvents: [.commandStarted])
+     *      client.enableNotifications(forEvents: [.commandSucceeded])
+     *
+     *  will result in only posting notifications for .commandSucceeded events.
+     */
+    public func enableNotifications(forEvents events: [MongoEvent] =
```
Comment (mbroadst) [inline Sources/MongoSwift/APM.swift]: nit: `enableCommandMonitoring` ? "Notifications" sounds pretty generic

File: Sources/MongoSwift/APM.swift [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+}
+
+/// An internal callback that will be set for "command started" events if the user 
+/// enables those notifications. This function generates a new `Notification` and posts 
+/// it to the default NotificationCenter for the app. 
+internal func commandStarted(_event: OpaquePointer?) {
+    guard let event = _event else { return }
+    let eventStruct = CommandStartedEvent(event)
+    let notification = Notification(name: .commandStarted, userInfo: ["event": eventStruct])
+    NotificationCenter.default.post(notification)
```
Comment (mbroadst) [inline Sources/MongoSwift/APM.swift]: we should explore ways to support non-default `NotificationCenter`s. One way might be to accept a `withNotificationCenter: NotificationCenter` that defaults to `NotificationCenter.default` in `enableCommandMonitoring` (I'm now starting to think that should be an imperative `monitorCommands([ ... ], withNotificationCenter: ... )` ? ).  Inside that method you can register a lambda that uses the provided `NotificationCenter`

File: Sources/MongoSwift/APM.swift [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+}
+
+/// An event published when a command starts. The event is stored under the key `event`
+/// in the `userInfo` property of `Notification`s posted under the name .commandStarted.
+public struct CommandStartedEvent {
+    /// An internal opaque pointer to a mongoc_apm_command_started_t 
+    internal let _event: OpaquePointer
+
+    /// An internal initializer for a CommandStartedEvent
+    internal init(_ event: OpaquePointer) {
```
Comment (mbroadst) [inline Sources/MongoSwift/APM.swift]: Rather than storing the opaque pointer here, I think you should "fill out" the struct with the data retrieved from the pointer. It might be unlikely that someone else deletes the pointer, but it _might_ happen.

Also what are the ownership rules for this pointer? Are we responsible for deleting it when we're done with it? If so, then we are presently leaking memory.

File: Sources/MongoSwift/APM.swift [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+     *  Notifications can only be enabled for a single NotificationCenter at a time.
+     *
+     *  Calling this function will reset all previously enabled events - i.e.
+     *  calling
+     *      client.enableNotifications(forEvents: [.commandStarted])
+     *      client.enableNotifications(forEvents: [.commandSucceeded])
+     *
+     *  will result in only posting notifications for .commandSucceeded events.
+     */
+    public func enableCommandMonitoring(
```
Comment (kmahar) [inline Sources/MongoSwift/APM.swift]: I'm wondering if we should rename this to also encompass SDAM monitoring - because the callbacks are set all at once for command monitoring + SDAM monitoring, we'll need the logic to create the `mongoc_apm_callbacks_t` in a single place. `enableApplicationPerformanceMonitoring`? or maybe just `enableMonitoring`?

File: Sources/MongoSwift/APM.swift [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+
+    /// The failure, represented as a MongoError.
+    let failure: MongoError
+
+    /// The client generated request id.
+    let requestId: Int64
+
+    /// The driver generated operation id. This is used to link events together such
+    /// as bulk write operations.
+    let operationId: Int64?
```
Comment (mbroadst) [inline Sources/MongoSwift/APM.swift]: is this really optional?

Comment (mbroadst) [reply Sources/MongoSwift/APM.swift]: ditto for `CommandSucceededEvent`

Comment (kmahar) [reply Sources/MongoSwift/APM.swift]: I took the optionality from the defined interfaces in the spec - it sounded like it would only be set for certain operation types.  but it appears that `libmongoc` always returns a value so I'll make it non-optional.

File: Sources/MongoSwift/APM.swift (hunk 5 above)
Comment (kmahar) [reply Sources/MongoSwift/APM.swift]: oops missed this comment before I merged, I'll rename it in the soon-to-come SDAM sketch PR
